plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/InvisibleCommand.java

```diff
@@ -73,87 +96,130 @@ else if (!scriptEntry.hasObject("target")
+        if (target == null || target.isFake) {
+            Debug.echoError(scriptEntry, "Must specify a valid target.");
+            return;
         }
-        if (target.isCitizensNPC()) {
-            NPC npc = target.getDenizenNPC().getCitizen();
-            if (!npc.hasTrait(InvisibleTrait.class)) {
-                npc.addTrait(InvisibleTrait.class);
-            }
-            InvisibleTrait trait = npc.getOrAddTrait(InvisibleTrait.class);
-            switch (Action.valueOf(state.asString().toUpperCase())) {
-                case FALSE:
-                    trait.setInvisible(false);
-                    break;
-                case TRUE:
-                    trait.setInvisible(true);
-                    break;
-                case TOGGLE:
-                    trait.toggle();
-                    break;
-            }
+        ElementTag state = scriptEntry.getElement("state");
+        List<PlayerTag> forPlayers = scriptEntry.argForPrefixList("for", PlayerTag.class, true);
+        if (scriptEntry.dbCallShouldDebug()) {
+            Debug.report(scriptEntry, getName(), target, state, db("for", forPlayers));
         }
-        else {
-            switch (Action.valueOf(state.asString().toUpperCase())) {
-                case FALSE:
-                    setInvisible(target, true);
-                    break;
-                case TRUE:
-                    setInvisible(target, false);
-                    break;
-                case TOGGLE:
-                    if (target.getBukkitEntity() instanceof ArmorStand) {
-                        setInvisible(target, !((ArmorStand) target.getBukkitEntity()).isVisible());
+        switch (state.asEnum(Action.class)) {
+            case TRUE:
+            case FALSE:
+                if (forPlayers == null) {
+                    setInvisible(target, state.asBoolean());
+                }
+                else {
+                    boolean invisible = state.asBoolean();
+                    for (PlayerTag player : forPlayers) {
+                        setInvisibleForPlayer(target, player, invisible);
                     }
-                    else if (target.getBukkitEntity() instanceof ItemFrame) {
-                        setInvisible(target, !((ItemFrame) target.getBukkitEntity()).isVisible());
+                }
+                break;
+            case TOGGLE:
+                if (forPlayers == null) {
+                    setInvisible(target, !isInvisible(target.getBukkitEntity(), null, false));
+                }
+                else {
+                    for (PlayerTag player : forPlayers) {
+                        setInvisibleForPlayer(target, player, !isInvisible(target.getBukkitEntity(), player.getUUID(), false));
                     }
-                    else {
-                        setInvisible(target, target.getLivingEntity().hasPotionEffect(PotionEffectType.INVISIBILITY));
+                }
+                break;
+            case RESET:

```

Comment:
Probably should handle `reset for:...`? (ie only reset the specified players)

Reply:
Added